### PR TITLE
feat: add gradmax details implementation and tests

### PR DESCRIPTION
**Pull request by maxencelebaron**
- Added weight normalization as described in the GradMax paper
- Added a custom ReLU activation with modified forward and backward passes to verify a key hypothesis from the paper
- Added corresponding unit tests

### CODE REVIEW

**Review by sylvchev — APPROVED** [review]
thanks @maxencelebaron this is a nice first PR 🎉 
I think the code is fine as is, we will need to improve the docstring of `normalize_optimal_updates` on the part that weere defined previously, but it is the scope of another PR.
For the experiment, did you have results (table/figure) that you could put in the discussion to see that we have the same results as the original code? As we will need this code quickly for upcoming deadline, you could add the results here (even if the PR is closed).

File: src/gromo/modules/growing_module.py [hunk 1]
```diff
@@ -2790,13 +2791,32 @@ def normalize_optimal_updates(
         and the extended_output_layer
         (so by s ** 2 / (std(extended_input_layer) * std(extended_output_layer)))
 
+        If normalization_type is "gradmax_normalization":
+        Let ``gradmax_scale`` be :math:`s` (default 1). Let :math:`c = s \\cdot \\text{mean}_i \\|W_i\\|`
+        where :math:`\\|W_i\\|` are the L2 (or Frobenius per channel) norms of the existing
+        slices of ``self.layer.weight`` along the input / fan-in axis (axis 1 for ``nn.Linear``).
```
Comment (TheoRudkiewicz) [inline src/gromo/modules/growing_module.py:2842]:
```suggestion
        If normalization_type is "gradmax_normalization":
        Let ``gradmax_scale`` be :math:`s` (default 1). Let :math:`c = s \cdot \text{mean}_i \|W_i\|`
        where :math:`\|W_i\|` are the L2 (or Frobenius per channel) norms of the existing
        slices of ``self.layer.weight`` along the input / fan-in axis (axis 1 for ``nn.Linear``).
```
Correct the math rendering.

File: src/gromo/modules/growing_module.py [hunk 2]
```diff
@@ -2808,8 +2828,74 @@ def normalize_optimal_updates(
...
+                if len(reduced_dims_existing) == 0:
+                    return
+
+                existing_norms = (layer_weight**2).sum(dim=reduced_dims_existing).sqrt()
+                if existing_norms.numel() == 0:
+                    return
+                target_norm = existing_norms.mean() * float(gradmax_scale)
+                if target_norm <= 0:
+                    return
+
```
Comment (TheoRudkiewicz) [inline src/gromo/modules/growing_module.py:2942]: I don't get why there is all these returns. The las one especially seems a bit useless.

File: src/gromo/utils/utils.py [hunk 3]
```diff
@@ -172,6 +204,8 @@ def activation_fn(fn_name: str) -> nn.Module:
     """
     known_activations = {
         "relu": nn.ReLU(),
+        "relu_fprime0_1": ReLUDerivativeOneAtZero(),
+        "relu_derivative_one_at_zero": ReLUDerivativeOneAtZero(),
```
Comment (stephane-rivaud) [inline src/gromo/utils/utils.py]: Why two different keys for the same class?

File: src/gromo/modules/growing_module.py (hunk 2 above)
Comment (stephane-rivaud) [reply src/gromo/modules/growing_module.py:2942]: I had hard times understanding the reason, but here is a summary:

---

`gradmax_normalization` has several guard clauses; each maps to a specific pathological shape/state and avoids crashes or silent NaNs:

- `len(reduced_dims_extension) == 0` / `len(reduced_dims_existing) == 0`: rank/pathology guard. We need “all axes except neuron axis (axis 1)” to compute per-neuron norms; if that set is empty (unexpected weight rank/layout), reduction is ill-defined. (think of (out_f=10, in_f=1) but encoded with a vector of size 10 instead of a matrix of size (10,1))
- `existing_norms.numel() == 0`: empty-axis guard (e.g. no slices along axis 1, shape like `(out, 0)`), so `mean()` would be invalid/NaN. (Think of empty layer that have a defined output dimension)
- `target_norm <= 0`: zero-signal guard. Even with `gradmax_scale > 0`, target can be 0 if all existing norms are 0. Current behavior is “no-op” (don’t collapse new weights to zero).

Alternatives and implications:
1) **Raise errors** instead of returning: stricter, surfaces misconfiguration early, but less robust for exploratory/partially initialized states.  
2) **Warn + return**: keeps robustness while improving debuggability (my preferred compromise).  
3) **Allow `target_norm == 0` and project to zero**: mathematically valid projection, but can wipe extension weights/eigenvalues and may be undesirable in practice.  
4) **Generalize axis handling** (derive neuron axis per layer type): better long-term than hardcoding axis 1, but more complexity and layer-specific plumbing.

File: tests/test_linear_growing_module.py [hunk 4]
```diff
@@ -3736,6 +3736,191 @@ def setup_layers_with_known_stds(
...
+            with self.subTest(case="gradmax_missing_extension_returns"):
+                _, layer_missing_ext = self.create_demo_layers_with_extension(
+                    include_eigenvalues=True,
+                    hidden_features=2,
+                    extension_size=2,
+                )
+                layer_missing_ext.extended_input_layer = None
+                layer_missing_ext.normalize_optimal_updates(
+                    normalization_type="gradmax_normalization"
+                )
```
Comment (stephane-rivaud) [inline tests/test_linear_growing_module.py]: I do not understand why a missing neuron extension is allowed when calling this function. This contradicts the safety measures taken when computing optimal update, where we would throw an error instead of silently letting this becoming a no-op.

File: src/gromo/utils/utils.py (hunk 3 above)
Comment (maxencelebaron) [reply src/gromo/utils/utils.py]: That must be a mistake on my part. I’m actually only using the second key

File: tests/test_linear_growing_module.py (hunk 4 above)
Comment (maxencelebaron) [reply tests/test_linear_growing_module.py]: You're right, silent no-ops are inconsistent with the rest of the codebase. Changed the return to raise ValueError and updated the test to use assertRaises. Thanks for catching this.

File: src/gromo/modules/growing_module.py (hunk 2 above)
Comment (maxencelebaron) [reply src/gromo/modules/growing_module.py:2942]: The explanation above is accurate. Each guard handles a specific edge case:
- len(reduced_dims_*) == 0: tensor has unexpected rank, can't compute per-neuron norms
- existing_norms.numel() == 0: empty axis, mean() would return NaN
- target_norm <= 0: all existing weights are zero, normalizing would collapse extension weights to zero

I'll add warnings before each return to make these cases visible during debugging, as suggested.

File: src/gromo/modules/growing_module.py [hunk 5]
```diff
@@ -2853,8 +2873,103 @@ def normalize_optimal_updates(
...
+                    extension_weight.mul_(scales.view(*reshape_shape))
+                    if self.eigenvalues_extension is not None:
+                        ev = self.eigenvalues_extension
+                        if ev.numel() != scales.numel():
+                            warnings.warn(
+                                "gradmax_normalization: eigenvalues_extension length "
+                                f"({ev.numel()}) does not match number of extension slices "
+                                f"({scales.numel()}); leaving eigenvalues unchanged.",
+                                UserWarning,
+                            )
```
Comment (TheoRudkiewicz) [inline src/gromo/modules/growing_module.py]: I don't get how is this possible ?

Comment (maxencelebaron) [reply src/gromo/modules/growing_module.py]: You're right, scales is built from extension_weight along axis 1, so it always has the same number of elements as eigenvalues_extension by construction. I'll remove the guard.

**Comment by TheoRudkiewicz** [comment]
@stephane-rivaud I think we can merge this. What do you think ?

**Review by TheoRudkiewicz — APPROVED** [review]

**Review by stephane-rivaud — APPROVED** [review]
Approving the changes.
